### DISCUSSION TITLE
Explanation of GlowTTS logp calculation

### DISCUSSION THREAD

**Discussion by p0p4k** [coqui-ai/TTS#1984] [General]
(Previous paper discussions : https://github.com/coqui-ai/TTS/discussions/1494, https://github.com/coqui-ai/TTS/discussions/1501, https://github.com/coqui-ai/TTS/discussions/1506, https://github.com/coqui-ai/TTS/discussions/1799)

Derivation for following GlowTTS lines of code :
```
'''
TTS/TTS/tts/models/glow_tts.py
lines 241 to 251
'''
# find the alignment path
with torch.no_grad():
    o_scale = torch.exp(-2 * o_log_scale)
    logp1 = torch.sum(-0.5 * math.log(2 * math.pi) - o_log_scale, [1]).unsqueeze(-1)  # [b, t, 1]
    logp2 = torch.matmul(o_scale.transpose(1, 2), -0.5 * (z**2))  # [b, t, d] x [b, d, t'] = [b, t, t']
    logp3 = torch.matmul((o_mean * o_scale).transpose(1, 2), z)  # [b, t, d] x [b, d, t'] = [b, t, t']
    logp4 = torch.sum(-0.5 * (o_mean**2) * o_scale, [1]).unsqueeze(-1)  # [b, t, 1]
    logp = logp1 + logp2 + logp3 + logp4  # [b, t, t']
    attn = maximum_path(logp, attn_mask.squeeze(1)).unsqueeze(1).detach()
y_mean, y_log_scale, o_attn_dur = self.compute_outputs(attn, o_mean, o_log_scale, x_mask)
attn = attn.squeeze(1).permute(0, 2, 1)
```

logp is the 2nd part of the summation in the following equation:
![image](https://user-images.githubusercontent.com/8834712/190890948-579f86d2-78b4-452c-a0e8-dd0ad73a662d.png)

![image](https://user-images.githubusercontent.com/8834712/190890430-4d342c4a-845c-4b6f-a705-e5c65144062f.png)

`attn` is computed using MAS algorithm, which I shall explain in next discussion.

**Comment by alifarrokh** [comment]
Brilliant!

**Comment by erogol** [comment]
Great post!!

**Reply by p0p4k** [reply]
Thanks man. 
PS. Been busy with other stuff, will get to reading more and adding minor PRs ASAP. Sorry for delay 😅
